Tests/GRPCTests/ClientConnectionBackoffTests.swift

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2019, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Foundation
+import GRPC
+import NIO
+import XCTest
+
+class ClientConnectionBackoffTests: XCTestCase {
+  let port = 8080
+
+  var client: EventLoopFuture<GRPCClientConnection>!
+  var server: EventLoopFuture<GRPCServer>!
+
+  var group: EventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+
+  override func tearDown() {
+    if let server = self.server {
+      XCTAssertNoThrow(try server.flatMap { $0.channel.close() }.wait())
+    }
+
+    // We don't always expect a client (since we deliberately timeout the connection in some cases).
+    if let client = try? self.client.wait() {
+      XCTAssertNoThrow(try client.channel.close().wait())
+    }
+
+    XCTAssertNoThrow(try self.group.syncShutdownGracefully())
+  }
+
+  func makeServer() throws -> EventLoopFuture<GRPCServer> {
+    return try GRPCServer.start(
+      hostname: "localhost",
+      port: self.port,
+      eventLoopGroup: self.group,
+      serviceProviders: [])
+  }
+
+  func makeClientConfiguration() -> GRPCClientConnection.Configuration {
+    return .init(
+      target: .hostAndPort("localhost", self.port),
+      eventLoopGroup: self.group,
+      connectionBackoff: ConnectionBackoff())
+  }
+
+  func makeClientConnection(
+    _ configuration: GRPCClientConnection.Configuration
+  ) -> EventLoopFuture<GRPCClientConnection> {
+    return GRPCClientConnection.start(configuration)
+  }
+
+  func testClientConnectionFailsWithNoBackoff() throws {
+    var configuration = self.makeClientConfiguration()
+    configuration.connectionBackoff = nil
+
+    self.client = self.makeClientConnection(configuration)
+    XCTAssertThrowsError(try self.client.wait())
+  }
+
+  func testClientEventuallyConnects() throws {
+    let clientConnected = self.expectation(description: "client connected")
+    let serverStarted = self.expectation(description: "server started")
+
+    // Start the client first.
+    self.client = self.makeClientConnection(self.makeClientConfiguration())
+    self.client.assertSuccess(fulfill: clientConnected)
+
+    self.server = try self.makeServer()
```

Comment:
Should we pause slightly before creating the server, to ensure that the back-off actually reconnects after a short delay?

Reply:
It shouldn't matter but there's no harm in doing, it certainly makes the test more "obvious".